cmd/gindoid/dstorage.go

```diff
@@ -31,59 +33,42 @@ func (ls *LocalStorage) Exists(target string) (bool, error) {
 }
 
 func (ls LocalStorage) Put(job DOIJob) error {
-	source := job.Source
-	target := job.Name
+	repopath := job.Source
+	jobname := job.Name
 	dReq := &job.Request
 
-	to := filepath.Join(ls.Path, target)
-	tmpDir := filepath.Join(to, tmpdir)
-	ls.prepDir(target, dReq.DOIInfo)
-	ds := ls.GetDataSource()
+	ls.prepDir(jobname, dReq.DOIInfo)
 
+	targetpath := filepath.Join(ls.Path, jobname)
 	preperrors := make([]string, 0, 5)
-
-	if out, err := ds.CloneRepository(source, tmpDir, &job.Key, ls.KnownHosts); err != nil {
-		log.WithFields(log.Fields{
-			"source": lpStorage,
-			"error":  err,
-			"out":    out,
-			"target": target,
-		}).Error("Repository cloning failed")
-		preperrors = append(preperrors, fmt.Sprintf("Failed to clone repository '%s': %s", source, err))
-	} else {
-		fSize, err := ls.zip(target)
-		if err != nil {
-			log.WithFields(log.Fields{
-				"source": lpStorage,
-				"error":  err,
-				"target": target,
-			}).Error("Could not zip the data")
-			preperrors = append(preperrors, fmt.Sprintf("Failed to create the zip file: %s", err))
-		}
-		// +1 to report something with small datasets
-		dReq.DOIInfo.FileSize = fSize/(1024*1000) + 1
+	zipsize, err := ls.cloneandzip(repopath, jobname, targetpath)
+	if err != nil {
+		// failed to clone and zip
+		// save the error for reporting and continue with the XML prep
+		preperrors = append(preperrors, err.Error())
 	}
-	ls.createIndexFile(target, dReq)
+	ls.createIndexFile(jobname, dReq)
+	dReq.DOIInfo.FileSize = zipsize/(1024*1000) + 1 // Proper size conversion to closest human-readable size
 
-	fp, err := os.Create(filepath.Join(to, "doi.xml"))
+	fp, err := os.Create(filepath.Join(targetpath, "doi.xml"))
 	if err != nil {
 		log.WithFields(log.Fields{
 			"source": lpStorage,
 			"error":  err,
-			"target": target,
+			"target": jobname,
 		}).Error("Could not create the metadata template")
 		preperrors = append(preperrors, fmt.Sprintf("Failed to create the XML metadata template: %s", err))
 	}
 	defer fp.Close()
```

Comment:
Not part of a change, but here the fp is defer closed even if the creation failed.